src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6845,75 +6845,66 @@ void reportBadDelegateParameter(BindingDiagnosticBag bag, MethodSymbol sourceInv
             }
         }
 
-        private void ReportNullabilityMismatchWithTargetDelegate(Location location, NamedTypeSymbol delegateType, UnboundLambda unboundLambda)
+        private void ReportNullabilityMismatchWithTargetDelegate(Location location, NamedTypeSymbol delegateType, BoundLambda lambda)
         {
-            if (!unboundLambda.HasExplicitlyTypedParameterList)
-            {
-                return;
-            }
+            MethodSymbol? targetInvokeMethod = delegateType.DelegateInvokeMethod;
+            LambdaSymbol sourceInvokeMethod = lambda.Symbol;
+            UnboundLambda unboundLambda = lambda.UnboundLambda;
 
-            var invoke = delegateType?.DelegateInvokeMethod;
-            if (invoke is null)
+            if (targetInvokeMethod is null ||
+                targetInvokeMethod.ParameterCount != sourceInvokeMethod.ParameterCount)
             {
                 return;
             }
 
-            Debug.Assert(delegateType is object);
-            if (unboundLambda.Syntax is LambdaExpressionSyntax lambdaSyntax)
+            // Parameter nullability is expected to match exactly. This corresponds to the behavior of initial binding.
+            //    Action<string> x = (object o) => { }; // error CS1661: Cannot convert lambda expression to delegate type 'Action<string>' because the parameter types do not match the delegate parameter types
+            //    Action<object> y = (object? o) => { }; // warning CS8622: Nullability of reference types in type of parameter 'o' of 'lambda expression' doesn't match the target delegate 'Action<object>'.
+            // https://github.com/dotnet/roslyn/issues/35564: Consider relaxing and allow implicit conversions of nullability (as we do for method group conversions).
+
+            if (lambda.Syntax is LambdaExpressionSyntax lambdaSyntax)
             {
                 int start = lambdaSyntax.SpanStart;
                 location = Location.Create(lambdaSyntax.SyntaxTree, new Text.TextSpan(start, lambdaSyntax.ArrowToken.Span.End - start));
             }
 
-            if (unboundLambda.HasExplicitReturnType(out _, out var returnType) &&
-                IsNullabilityMismatch(invoke.ReturnTypeWithAnnotations, returnType, requireIdentity: true))
+            if (SourceMemberContainerTypeSymbol.CheckValidNullableMethodOverride(
+                compilation,
+                targetInvokeMethod,
+                sourceInvokeMethod,
+                new BindingDiagnosticBag(Diagnostics),
```

Comment:
It is not part of your change, but I am surprised that `BindingDiagnosticBag` is needed here.

It feels like we could at least reuse the instance between the two calls.

Reply:
It looks like BindingDiagnosticBag is still created again for each call.